File: ifad/ifad_scrape.py
################################################################################
# ifad/ifad-scrape.py                                                          #
#                                                                              #
# Copyright 2021 Evans Policy Analysis and Research Group (EPAR).              #
#                                                                              #
# This project is licensed under the 3-Clause BSD License. Please see the      #
# license.txt file for more information.                                       #
#                                                                              #
# This script crawls the International Fund for Agricultural Development       #
# project pages to create a database of specified information                  #
################################################################################

__copyright__ = """
Copyright 2021 Evans Policy Analysis and Research Group (EPAR).
"""
__license__ = """
This project is licensed under the 3-Clause BSD License. Please see the 
license.txt file for more information.
"""

# Imports
import csv
import html
import pandas as pd
import re
import requests
import sys
import time
import unidecode
from bs4 import BeautifulSoup

# Constants
DEBUG = False if len(sys.argv) == 1 else sys.argv[1] == "True"
BASE_URL = 'https://www.ifad.org/en/web/operations/projects-and-programmes?mode=search'
TABS = [1,2,3]
PROJECT_URL = 'https://www.ifad.org/en/web/operations/-/project/'
OUTPUT_FILE = './data/ifad_data_debug.csv' if DEBUG else './data/ifad_data.csv'
IFI_COUNTRIES = {
    'Angola': 'Angola',
    'Benin': 'Benin',
    'Botswana': 'Botswana',
    'Burkina Faso': 'Burkina Faso',
    'Burundi': 'Burundi',
    'Cameroon': 'Cameroon',
    'Cabo Verde': 'Verde',
    'Central African Republic': 'Central African Republic',
    'Chad': 'Chad',
    'Comoros': 'Comoros',
    "Côte d'Ivoire": "Côte d'Ivoire",
    'Democratic Republic of the Congo': 'Democratic Republic of the Congo',
    'Equatorial Guinea': 'Equatorial Guinea',
    'Eritrea': 'Eritrea',
    'Eswatini': 'Eswatini',
    'Ethiopia': 'Ethiopia',
    'Gabon': 'Gabon',
    'Gambia': 'Gambia',
    'Gambia (The)': 'Gambia',
    'Ghana': 'Ghana',
    'Guinea': 'Guinea',
    'Guinea-Bissau': 'Guinea-Bissau',
    'Kenya': 'Kenya',
    'Lesotho': 'Lesotho',
    'Liberia': 'Liberia',
    'Madagascar': 'Madagascar',
    'Malawi': 'Malawi',
    'Mali': 'Mali',
    'Mauritania': 'Mauritania',
    'Mauritius': 'Mauritius',
    'Mozambique': 'Mozambique',
    'Namibia': 'Namibia',
    'Niger': 'Niger',
    'Nigeria': 'Nigeria',
    'Republic of Congo': 'Republic of the Congo',
    'Rwanda': 'Rwanda',
    'Sao Tome and Principe': 'Sao Tome and Principe',
    'Senegal': 'Senegal',
    'Seychelles': 'Seychelles',
    'Sierra Leone': 'Sierra Leone',
    'South Africa': 'South Africa',
    'South Sudan': 'South Sudan',
    'Tanzania': 'Tanzania',
    'United Republic of Tanzania': 'Tanzania',
    'Togo': 'Togo',
    'Uganda': 'Uganda',
    'Zambia': 'Zambia',
    'Zimbabwe' : 'Zimbabwe'
}

def get_html(url):
    attempts = 0
    while(attempts < 20):
        try:
            attempts += 1
            response = requests.get(url)
            clean_html = html.unescape(response.text)
            return BeautifulSoup(clean_html, 'html.parser')
        except (Exception) as e:
            print('Failed to download webpage, trying again')
            time.sleep(5)
    return None    

def get_proj_ids(url, tabs):
    """
    This function takes the BASE_URL and TABS to search
    and returns the list of project IDs to scrape
    """
    soup = get_html(url)
    projects = list()
    for i in tabs:
        # These are lists of HTML tags. use <element>.text to get to the actual text
        countries = soup.select('div.tab' + str(i) + ' div.project-info-container div.col-md-3')
        proj_ids = soup.select('div.tab' + str(i) + ' div.project-info-container div.col-md-2')
        del proj_ids[1::2] #project dates also match the col-md-2 filter; remove them by dropping every other match
        assert(len(proj_ids) == len(countries))
        # Merge IDs and countries into a single list of tuples
        id_and_country = tuple(zip(proj_ids, countries))
        # Filter projects for IFI countries "list(filter(lambda...))", then take only the project IDs from the resulting list (a_tuple[0].text)
        # Concepts: "filtering with lambdas" and "list comprehensions"
        relevant_ids = [a_tuple[0].text for a_tuple in (list(filter(lambda t : (t[1].text in IFI_COUNTRIES.keys()), id_and_country)))]
        projects.extend(relevant_ids)
    return projects

# Manual scraping method that finds param:to_find in param:soup and places its value in param:data
def manual_scrape(soup, data, to_find):
    try:
        ret = soup.find('dt', text=re.compile(r"\s*" + to_find + "\s*")).findNext().text.strip()
        data[to_find] = ret
        return ret
    except Exception as e:
        print(e)
        print('\t{0} not found'.format(to_find))

# Main
projects = get_proj_ids(BASE_URL, TABS) if not DEBUG else ['2000003362', '2000001936']
scraped_data = []

for project_id in projects:
    data = {}
    url = PROJECT_URL + project_id
    print('Scraping {0}'.format(url))
    soup = get_html(url)
    if soup == None:
        next
    data['Project ID'] = project_id
    manual_scrape(soup, data, 'Country')
    data['Project Title'] = soup.select("h1[class!=\"hide-accessible\"]")[0].text
    manual_scrape(soup, data, 'Total Project Cost')
    data['Status'] = soup.select('dd.project-status > span')[0].text[8:]
    manual_scrape(soup, data, 'Approval Date')
    manual_scrape(soup, data, 'Duration')
    manual_scrape(soup, data, 'IFAD Financing')
    manual_scrape(soup, data, 'Financing Gap')

    # Handle multiple international funders
    int_funders = ''
    f = soup.find(text='Co-financiers (International)')
    while f != None and 'project-row-text' in f.findNext()['class']:
        int_funders += f.findNext().text.strip() + '); '
        f = f.findNext()
    int_funders = int_funders[:-2].replace('US$', '(US$') # String cleanup
    
    # Handle multiple domestic funders
    dom_funders = ''
    f = soup.find(text='Co-financiers (Domestic)')
    while f != None and 'project-row-text' in f.findNext()['class']:
        dom_funders += f.findNext().text.strip() + '); '
        f = f.findNext()
    dom_funders = dom_funders[:-2].replace('US$', '(US$') # String cleanup
    
    # Add international and domestic funders to the lists
    if len(int_funders) > 0:
        data['Co-financiers (International)'] = int_funders
    if len(dom_funders) > 0:
        data['Co-financiers (Domestic)'] = dom_funders

    manual_scrape(soup, data, 'Financing terms')
    manual_scrape(soup, data, 'Sector')
    contact_name = manual_scrape(soup, data, 'Project Contact')
    if contact_name != None and soup.find(text=contact_name) != None:
        data['Contact Details'] = soup.find(text=contact_name).parent['href'][7:]

    for key in data.keys():
        # Remove special characters, but not from country names
        if key != 'Country':
            data[key] = unidecode.unidecode(data[key].strip()).strip() 
        # Translate country names into IFI format
        else:
            data[key] = IFI_COUNTRIES[data[key]]
    
    scraped_data.append(data)
    
    # Print the scraped data
    [print('\t{0}: {1}'.format(key, value)) for key, value in data.items()] 
    print()

# Export into excel file
print("Creating excel file '{0}' with scraped data".format(OUTPUT_FILE))
df = pd.DataFrame.from_records(scraped_data)

# Don't fail because the output file was open
while True:
    try:
        df.to_csv(open(OUTPUT_FILE, 'w'), index=False, line_terminator='\n', na_rep='NA')
        break
    except Exception as e:
        print("Failed to write to CSV file. Please make sure that 1) file is closed, and 2) you are running this script from the 411-IFI-Aid/ folder.")
        time.sleep(5)

print('All done!')

File: ifad/ifad-scrape.R
#!/usr/bin/env Rscript

################################################################################
# ifad/ifad-scrape.R                                                           #
#                                                                              #
# Copyright 2021 Evans Policy Analysis and Research Group (EPAR).              #
#                                                                              #
# This project is licensed under the 3-Clause BSD License. Please see the      #
# license.txt file for more information.                                       #
#                                                                              #
# this script crawls International Fund for Agricultural Development project   #
# pages to create a database of specified information                          #
#                                                                              #
#                                                                              #
# inputs: none                                                                 #
# outputs: a csv file containing project numbers, description, planned         #
#          completion date, dac sector codes, etc                              #
################################################################################

# IMPORTS #
# install.packages("tidyverse")
# install.packages("xml2")
# install.packages("magrittr")
# install.packages("rvest")
# install.packages("sjmisc")
#install.packages("R.utils",repos = "http://cran.us.r-project.org")
library(magrittr)
library(readxl)
library(R.utils)
library(rvest)
library(sjmisc)
library(tidyverse)
library(xml2)

# CONSTANTS AND OUTPUT #
DEBUG <- TRUE
BASE_URL <- "https://www.ifad.org/en/web/operations/projects-and-programmes?mode=search"
PROJECT_URL <- "https://www.ifad.int/en/web/operations/-/project/"
OUTPUT_FILE <- if(DEBUG) "../data/ifad_test.csv" else "../data/ifad_data.csv"
DAC_FILE <- "../DAC-CRS-CODES.xls"
IFI_COUNTRIES <- c("Angola", "Benin", "Botswana", "Burkina Faso", "Burundi", "Cameroon",
"Cabo Verde","Central African Republic","Chad","Comoros","Côte d'Ivoire","Democratic Republic of the Congo",
"Equatorial Guinea","Eritrea","Eswatini","Ethiopia","Gabon","Gambia","Gambia (The)", "Ghana","Guinea",
"Guinea-Bissau","Kenya","Lesotho","Liberia","Madagascar","Malawi","Mali","Mauritania",
"Mauritius","Mozambique","Namibia","Niger","Nigeria","Republic of Congo","Rwanda",
"Sao Tome and Principe","Senegal","Seychelles","Sierra Leone","South Africa","South Sudan",
"Tanzania", "United Republic of Tanzania", "Togo","Uganda","Zambia","Zimbabwe")

#sink("ifad-output.txt", split=TRUE, append = FALSE)

# FUNCTIONS #
#Scrape
get_html <- function(url) {
  tryCatch(
    read_html(url),
    error = function(e) {
      cat("Request failed, retrying.\n")
      Sys.sleep(5) 
      NULL
    }
  )
}

# File write
write_file <- function(df, filename) {
  tryCatch(
    {
      df %>% write.csv( 
        file = filename, 
        quote = TRUE)
      TRUE
    },
    error = function(e) {
      cat("writing to file failed, retrying.\n")
      print(e)
      FALSE
    }
  )
}

#Looks up the DAC or DAC5 code in the DAC-CRS-CODES excel file
get_dac5_desc <- function(code){
  if(is.null(code) || code == "N/A") return("N/A")
  code <- strtoi(code)
  desc <- if(code < 1000)
    dac_df$DESCRIPTION[which(dac_df$`DAC 5 CODE` == code)] 
  else
    dac_df$DESCRIPTION[which(dac_df$`concatenate` == code)]
  return(if(!is.null(desc)) desc else "N/A")
}

#Scrape project IDs out of current tab
scrape_rows <- function(page, tab){
  #Get each project's row element
  rows <- page %>% html_nodes(paste("div.tab", tab, sep="")) %>% 
    html_nodes("div.container > div.row") %>% 
    html_nodes("div.project-info-paragraph > a") %>% 
    html_nodes("div.col-md")

  #Filter to projects from relevant countries
  countries <- rows %>% html_nodes("div.col-md-3") %>% html_text()
  is_relevant_country <- countries %in% IFI_COUNTRIES
  
  #Grab project IDs
  ids <- rows %>% html_nodes("div.col-md-2") %>% html_text()
  #Filter out dates
  ids <- ids[!grepl("\\s", ids)]
  #Filter our irrelevant countries
  ids <- ids[is_relevant_country]
  return(ids)
}

# MAIN #
#Read the DAC description excel file
dac_df <- read_excel(DAC_FILE, sheet=12, skip=2)
page <- get_html(BASE_URL)

ids <- scrape_rows(page, 1)
ids <- append(ids, scrape_rows(page, 2))
ids <- append(ids, scrape_rows(page, 3))
print(ids)

# create a dataframe to hold the scraped data.
data <- data.frame(matrix(ncol = 18, nrow = 0))
names(data) <- c(
  "Project ID",
  "Country",
  "Project Title",
  "Description",  
  "Commitment in U.A.",
  "Status",
  "Start Date",
  "Closing Date",
  "Project Duration",
  "Source of Financing",
  "Sovereign",
  "Sector",
  "DAC Sector Code",
  "DAC5 Code",
  "DAC5 Description",
  "Detailed Description",
  "Contact Name",
  "Contact Email"
)

# for each id in provided file
for(id in ids) {
  if(is.null(id) || is.na(id)) next
  
  start <- Sys.time()
  "scraping project" %>% paste(id, "\n", sep = " ") %>% cat()
  page  <- NULL
  count <- 0
  
  # using a loop to retry on failed attempts just in case there is an http
  # request error. only retry 20 times in case there is a bad id that doesn't
  # work (it happened, don't remove the loop limiter)
  while(length(page) == 0 && count < 20){
    page <- get_html(paste(PROJECT_URL, id, sep=""))
    count <- count + 1
  }
  # if the loop ended because of the loop limiter skip to next id
  if(length(page) == 0) next
  processing_time <- Sys.time()
  ##############################
  # parse the web page results #
  ##############################
  # PULL INFO FROM SITE
  
  
  if(DEBUG) print(paste(id, country, project, commitment, approval_date, completion_date, duration, funding, status, sector, sov, dac, dac5, dac5_desc, dac5_desc_detailed, contact_name, contact_email, sep="; "))
  # add parsed data into the main data frame
  data[nrow(data) + 1,] <- c(id, country, project, desc, commitment, approval_date, completion_date, duration, funding, status, sector, sov, dac, dac5, dac5_desc, dac5_desc_detailed, contact_name, contact_email)
  
  # calculate elapsed time for request
  elapsed <- Sys.time() - start
  # Respect the robots.txt - only do a request every 10 seconds
  printf("Download time: %.2f // Processing time: %.2fs // Sleep time: %.2fs\n", processing_time - start, elapsed - processing_time + start, 10 - elapsed)
  if(elapsed < 10) {
    Sys.sleep(10-elapsed)
  }
}

# Write to disk
written <- FALSE
count <- 0
while(!written && count < 20){
  written <- write_file(data, OUTPUT_FILE)
  count <- count + 1
}
closeAllConnections()
print("All done!")